File: codingTest/3week-assignment/day3/B1654.py
#https://www.acmicpc.net/problem/1654
#랜선자르기


k, n = map(int, input().split())
lan = [int(input()) for _ in range(k)]

max_length = max(lan)
min_length = 1  # 최소 길이를 0이 아닌 1로 설정

while min_length < max_length:
    mid = (max_length + min_length) // 2
    count = sum(lan_length // mid for lan_length in lan)

    if count < n:
        max_length = mid
    else:
        min_length = mid + 1

print(min_length - 1)

File: java-version/src/week3/day3/B1654.java
package week3.day3;
//https://www.acmicpc.net/problem/1654
//랜선 자르기

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class B1654 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st =new StringTokenizer(br.readLine());
        int k = Integer.parseInt(st.nextToken());
        int n = Integer.parseInt(st.nextToken());

        int[] lan = new int[k];
        int max=0;
        int min=0;

        for (int i = 0; i < k; i++) {
            lan[i] = Integer.parseInt(br.readLine());
            max = Math.max(lan[i], max);
        }


        while (min < max) {
            int mid = (max + min)/2;
            long count = 0;



            for (int lan_len : lan) {
                count+= lan_len/mid;
            }



            if(count < n) {
                max = mid;
            }
            else {
                min = mid + 1;
            }


        }

        System.out.println(min - 1);

    }
}
